File: java/src/io/github/maximerihouey/Benchmark.java
package io.github.maximerihouey;

import java.util.concurrent.ThreadLocalRandom;
import java.util.stream.*;
import java.io.*;

/**
 * Created by maxime on 30/10/16.
 */
public class Benchmark {

    public static int[] range(int n, int startIndex, int stepSize){
        int[] array = new int[n];
        for(int i=0; i < n; i++){
            array[i] = startIndex + i * stepSize;
        }
        return array;
    }

    public static void fisherYatesShuffle(int[] array){
        for(int i=0; i <= array.length-2; i++){
            int j = ThreadLocalRandom.current().nextInt(array.length - 1);
            Quickselect.switcharoo(array, i, j);
        }
    }

    public static long[] timeExecution(int[] example, int k, int nbExampleBySize){
        long[] times = new long[nbExampleBySize];
        for(int i=0; i < nbExampleBySize; i++){
            fisherYatesShuffle(example);
            long start = System.nanoTime();
            Quickselect.quickselect(example, k);
            times[i] = System.nanoTime() - start;
        }
        return times;
    }

    public static void main(String args[]){
        // FileWriter
        File file = new File("../benchmark_results.txt");
        try {
            BufferedWriter bw = new BufferedWriter(new FileWriter(file));

            int[] exampleSizes = {50, 75, 100, 250, 500, 750, 1000, 1500, 2000, 3000, 4000, 5000, 7500, 9000, 10000, 15000, 17500, 20000, 25000};
            int nbExampleBySize = 250;

            for(int i=0; i < exampleSizes.length; i++){
                int exampleSize = exampleSizes[i];
                System.out.println(exampleSize);
                int[] example = range(exampleSize, 1, 1);
                int medianIndex = exampleSize/2;
                long[] times = timeExecution(example, medianIndex, nbExampleBySize);
                long sum = LongStream.of(times).sum();
                bw.write(exampleSize + " " + sum/(times.length * Math.pow(10, 9)) + "\n");
            }

            bw.close();
        } catch (IOException exception) {
            System.out.println("Unable to open file");
        }
    }
}

File: C/benchmark.c
#include <stdio.h>
#include <stdlib.h>
#include "quickselect.h"
#include <time.h>
#include <unistd.h>

int* range(int n, int startIndex, int stepSize){
    int *array = (int *)malloc(sizeof(int) * n);
    int i;
    for(i=0; i < n; i = i + 1 ){
        array[i] = startIndex + i * stepSize;
    }
    return array;
}

void fisherYatesShuffle(int *array, int array_size){
    int i, j;
    for(i=0; i < array_size-1; i = i + 1 ){
        j = (rand() % (array_size));
        switcharoo(array, i, j);
    }
}

long* timeExecution(int *example, int example_size, int k, int nbExampleBySize){
    long *times = (long *)malloc(sizeof(long) * nbExampleBySize);
    int i;
    struct timespec start, stop;
    for(i=0; i < nbExampleBySize; i = i + 1 ){
        fisherYatesShuffle(example, example_size);
        clock_gettime(CLOCK_REALTIME, &start);
        quickselect(example, example_size, k);
        clock_gettime(CLOCK_REALTIME, &stop);
        times[i] = (stop.tv_nsec - start.tv_nsec);
    }
    return times;
}

double mean(long *array, int array_size){
    int i;
    long somme = 0;
    for(i=0; i < array_size; i = i + 1 ){
        somme += array[i];
    }
    return (double) somme / (double) array_size;
}

int main(void)
{
    int i;
    const int exampleSizes[] = {50, 75, 100, 250, 500, 750, 1000, 1500, 2000, 3000, 4000, 5000, 7500, 9000, 10000, 15000, 17500, 20000, 25000};
    int nbExampleBySize = 250;
    int *example;
    long temp_somme;
    int example_size;
    double mean_duration;
    FILE *f = fopen("benchmark_results.txt", "w");

    for(i=0; i < 19; i = i + 1 ){
        example_size = exampleSizes[i];
        example = range(example_size, 1, 1);
        mean_duration = mean(timeExecution(example, example_size, example_size/2, nbExampleBySize), nbExampleBySize);

        printf("%i\n", example_size);
        fprintf(f, "%i %.10f\n", example_size, mean_duration / 1000000000L);
    }
    fclose(f);
}



